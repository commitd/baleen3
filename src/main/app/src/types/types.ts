/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { Annot8ComponentDescriptor } from './base-types'
import {
  Annot8ComponentInfo,
  Measurement,
  PipelineTemplate,
  SettingsSchema,
  WithDescription,
  WithName,
  ErrorConfiguration,
} from './server-types'

/**
 * Index of metrics measurements
 */
export interface MeasurementContainer {
  [key: string]: Measurement[]
}

/**
 * Holds all the classes that have metrics associated with them, and their associated measurements
 */
export interface MetricsContainer {
  [key: string]: MeasurementContainer
}

export interface SettingsSchemaExtended extends SettingsSchema {
  /**
   * String version of json uiSchema
   *
   */
  uiSchema: string
}

export interface ComponentInfo extends Annot8ComponentInfo {
  name: string
}

/**
 * Map of server components
 */
export interface ComponentMap {
  [key: string]: ComponentInfo
}

/**
 * Map of settings schema
 */
export interface SettingsMap {
  [key: string]: SettingsSchema
}

/**
 * The full set of customisation details for the server.
 */
export interface ServerDetails {
  /**
   * Index by class name of ComponentInfo that represent the processors available on the server that can be used in pipelines.
   */
  processors: ComponentMap
  /**
   * Index by class name of ComponentInfo that represent the sources available on the server that can be used in pipelines.
   */
  sources: ComponentMap
  /**
   * Index by class name of SettingsSchema that are the settings classes of all sources and processors available on the server that can be used in pipelines.
   */
  settings: SettingsMap
  /**
   * Index by class name of ComponentInfo that represent the orderers available on the server that can be used to order pipelines.
   * <p>
   * NB this is currently generated from a strings to simulate a ComponentInfos
   */
  orderers: ComponentMap
  /**
   * Pipeline templates that can be used to start the process of creating a pipeline
   */
  templates: PipelineTemplate[]
}

export interface ComponentMetadata {
  id: string
  descriptor: Annot8ComponentDescriptor
  info: ComponentInfo | undefined
  schema: SettingsSchema | undefined
  valid: boolean
  configured: boolean
}

/**
 * Mapping of Pipeline descriptor to include extra metadata
 *
 */
export interface PipelineHeaderDescriptor extends WithName, WithDescription {}

/**
 * Mapping of Pipeline descriptor to include extra metadata
 *
 */
export interface PipelineViewDescriptor extends PipelineHeaderDescriptor {
  sourceOrder: string[]
  processorOrder: string[]
  components: Record<string, ComponentMetadata>
  errorConfiguration: ErrorConfiguration
}

/**
 * Utility tracking interface to help maintain the UI key state
 */
export interface PipelineEditDescriptor extends PipelineViewDescriptor {
  nameValid: boolean
  orderer: string
  errors: string[]
}

export type ComponentType = 'Processor' | 'Source'

export const NO_SETTINGS: 'io.annot8.api.settings.NoSettings' =
  'io.annot8.api.settings.NoSettings'

export const NO_OP_ORDERER: 'io.annot8.api.pipelines.NoOpOrderer' =
  'io.annot8.api.pipelines.NoOpOrderer'
