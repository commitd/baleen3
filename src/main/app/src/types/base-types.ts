/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
// We remove the generics for the Settings type from these as we never know what they are
// so do not help with any compile time errors

export interface ComponentDescriptor {
  name: string
  settings?: object
}

export interface Annot8ComponentDescriptor {
  [_class: string]: ComponentDescriptor
}

// Types required to complete generated server types
// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars-experimental
export interface SourceDescriptor extends Annot8ComponentDescriptor {}
// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unused-vars-experimental
export interface ProcessorDescriptor extends Annot8ComponentDescriptor {}

/**
 * Interface for describing known metrics, to give better display.
 */
export interface MetricMetadata {
  name?: string
  description?: string
  units?: string
  transform?: (v: number) => number
}

/**
 * Provided stacktrace type to correct optional properties
 */
export interface StackTrace {
  classLoaderName?: string
  className: string
  fileName?: string
  lineNumber: number
  methodName: string
  moduleName?: string
  moduleVersion?: string
  nativeMethod?: boolean
}

/**
 * Provided throwable type to correct optional properties
 */
export interface Throwable {
  cause?: Throwable
  localizedMessage: string
  message: string
  stackTrace: StackTrace[]
  suppressed?: Throwable[]
}
