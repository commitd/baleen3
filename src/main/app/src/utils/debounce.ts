/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
/* eslint-disable @typescript-eslint/no-explicit-any */
export type Method = (...args: any[]) => void

export function debounce<F extends Method>(
  func: F,
  wait = 200
): (this: ThisParameterType<F>, ...args: Parameters<F>) => void {
  let timeoutId: ReturnType<typeof setTimeout> | undefined

  return function (this: ThisParameterType<F>, ...args: Parameters<F>): void {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const context = this

    if (timeoutId !== undefined) {
      clearTimeout(timeoutId)
    }
    timeoutId = setTimeout(() => {
      timeoutId = undefined
      func.apply(context, args)
    }, wait)
  }
}
