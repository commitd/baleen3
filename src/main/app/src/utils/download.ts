/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const downloadJson = (name: string, data: any): void => {
  download(
    name + '.json',
    'data:text/json;charset=utf-8,',
    JSON.stringify(data, null, 2)
  )
}

export const download = (
  name: string,
  mediaType: string,
  data: string
): void => {
  const dataStr = mediaType + encodeURIComponent(data)
  const downloadAnchorNode = document.createElement('a')
  downloadAnchorNode.setAttribute('href', dataStr)
  downloadAnchorNode.setAttribute('download', name)
  document.body.appendChild(downloadAnchorNode) // required for firefox
  downloadAnchorNode.click()
  downloadAnchorNode.remove()
}
