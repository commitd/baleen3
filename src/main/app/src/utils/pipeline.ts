/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import Ajv from 'ajv'
import { v4 } from 'uuid'
import {
  Annot8ComponentDescriptor,
  ComponentInfo,
  ComponentMap,
  ComponentMetadata,
  PipelineDescriptor,
  PipelineEditDescriptor,
  PipelineViewDescriptor,
  ServerDetails,
  SettingsMap,
  SettingsSchema,
} from '../types'

const ajv = new Ajv({ useDefaults: true })
// eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-return
const uuid = (): string => v4()

export const isSettingsValid = (
  schema?: SettingsSchema,
  settings?: object
): boolean => {
  if (schema === undefined) {
    // No settings schema, assume valid
    return true
  }

  try {
    const jsonSchema = JSON.parse(schema.jsonSchema) as object
    const validator = ajv.compile(jsonSchema)
    const valid = validator(settings) as boolean
    if (!valid) {
      let message = 'JSON not valid:'
      if (!(validator.errors == null)) {
        validator.errors.forEach((error: Ajv.ErrorObject): void => {
          message += error.message
        })
      }
      console.debug(message)
    }
    return valid
  } catch (e) {
    console.debug(e)
    return false
  }
}

export const defaultSettings = (
  schema?: SettingsSchema
): object | undefined => {
  if (schema === undefined) {
    // No settings schema, assume no settings
    return undefined
  }

  try {
    const jsonSchema = JSON.parse(schema.jsonSchema) as object
    const validator = ajv.compile(jsonSchema)
    const defaultData = {}
    const valid = validator({}) as boolean
    if (valid) {
      return defaultData
    }
  } catch (e) {
    console.debug(e)
  }
  return undefined
}

export const isAllComponentsValid = (
  components: ComponentMetadata[]
): boolean => {
  return components.find((c) => !c.valid) === undefined
}

export const isPipelineValid = (
  nameValid: boolean,
  pipeline: Readonly<PipelineViewDescriptor>
): string[] => {
  const errors = []
  if (!nameValid) {
    errors.push('The pipeline name must be unique')
  }
  if (pipeline.sourceOrder.length === 0) {
    errors.push('A pipeline must have at least one source')
  }
  if (pipeline.processorOrder.length === 0) {
    errors.push('A pipeline must have at least one processor')
  }
  if (!isAllComponentsValid(Object.values(pipeline.components))) {
    errors.push('All components must be valid')
  }

  return errors
}

export const createComponent =
  (settings: SettingsMap, types: ComponentMap, configured = true) =>
  (descriptor: Annot8ComponentDescriptor): ComponentMetadata => {
    const id = uuid()
    const type = Object.keys(descriptor)[0]
    const info: ComponentInfo | undefined = types[`${type}`]
    const schema: SettingsSchema | undefined =
      info !== undefined && info.settingsClass !== undefined
        ? settings[info.settingsClass]
        : undefined

    if (schema !== undefined && descriptor[`${type}`].settings === undefined) {
      descriptor[`${type}`].settings = defaultSettings(schema)
    }

    return {
      id,
      descriptor,
      info,
      schema,
      valid:
        info !== undefined &&
        isSettingsValid(schema, descriptor[`${type}`].settings),
      configured,
    }
  }

export const createPipelineViewDescriptor = (
  pipeline: PipelineDescriptor,
  serverDetails: ServerDetails
): PipelineViewDescriptor => {
  const components: { [key: string]: ComponentMetadata } = {}
  const sourceOrder: string[] = []
  const processorOrder: string[] = []

  pipeline.sources
    .map(createComponent(serverDetails.settings, serverDetails.sources))
    .forEach((component) => {
      components[component.id] = component
      sourceOrder.push(component.id)
    })

  pipeline.processors
    .map(createComponent(serverDetails.settings, serverDetails.processors))
    .forEach((component) => {
      components[component.id] = component
      processorOrder.push(component.id)
    })

  return {
    name: pipeline.name,
    description: pipeline.description,
    sourceOrder,
    processorOrder,
    components,
    errorConfiguration: pipeline.errorConfiguration,
  }
}

export const createPipelineEditDescriptor = (
  pipeline: PipelineDescriptor,
  orderer: string,
  usedNames: string[],
  serverDetails: ServerDetails
): PipelineEditDescriptor => {
  const pipelineView = createPipelineViewDescriptor(pipeline, serverDetails)

  const nameValid = !usedNames.includes(pipeline.name)

  return {
    ...pipelineView,
    orderer,
    nameValid,
    errors: isPipelineValid(nameValid, pipelineView),
  }
}
