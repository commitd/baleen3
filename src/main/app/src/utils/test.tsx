/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { StylesProvider } from '@material-ui/styles'
import {
  createHistory,
  createMemorySource,
  History,
  LocationProvider,
} from '@reach/router'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { GenerateId } from 'jss'
import React from 'react'
import { ThemeProvider } from '../components/ThemeProvider/ThemeProvider'

const generateClassName: GenerateId = (rule, styleSheet) => {
  const prefix =
    styleSheet === undefined || styleSheet.options.classNamePrefix === undefined
      ? ''
      : styleSheet.options.classNamePrefix
  return `${prefix}-${rule.key}`
}

const LightTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="light">{children}</ThemeProvider>
  </StylesProvider>
)

const DarkTheme: React.FC<{ readonly children?: React.ReactNode }> = ({
  children,
}) => (
  <StylesProvider generateClassName={generateClassName}>
    <ThemeProvider choice="dark">{children}</ThemeProvider>
  </StylesProvider>
)

export const renderPlain = render

export const renderLight = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: LightTheme, ...options })

export const renderDark = (
  ui: Readonly<React.ReactElement>,
  options?: Omit<RenderOptions, 'queries'>
): RenderResult => render(ui, { wrapper: DarkTheme, ...options })

export function renderWithRouter(
  ui: Readonly<React.ReactElement>,
  route = '/',
  options?: Omit<RenderOptions, 'queries'>,
  delegate = renderLight
): RenderResult & { history: History } {
  const history = createHistory(createMemorySource(route))
  return {
    ...delegate(
      <LocationProvider history={history}>{ui}</LocationProvider>,
      options
    ),
    history,
  }
}
// re-export everything
export * from '@testing-library/react'
// override render method
export { renderLight as render }
export { userEvent }
