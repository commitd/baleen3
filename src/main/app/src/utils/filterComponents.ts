/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import { ComponentInfo, ComponentMap } from '../types'
import { parseSearch, standardize } from './text'

export const filterComponents = (
  map: ComponentMap,
  search: string
): Array<ComponentInfo> => {
  const values = Object.values(map)
  if (search.trim() === '') {
    return values
  }

  const searchTerms = parseSearch(search)

  return values.filter(
    (value) =>
      filterBy('artifact', value, searchTerms) ||
      filterBy('description', value, searchTerms) ||
      filterByName(value, searchTerms) ||
      filterByTags(value, searchTerms)
  )
}

function filterBy(
  key: keyof Omit<ComponentInfo, 'tags'>,
  value: ComponentInfo,
  terms: string[]
): boolean {
  // false positive security/detect-object-injection
  // eslint-disable-next-line
  const property = value[key]
  if (property === undefined) {
    return false
  }
  return terms.some((term) => standardize(property).includes(term))
}

function filterByName(value: ComponentInfo, terms: string[]): boolean {
  const property = standardize(value.name)
  return terms.some((term) => property.includes(term))
}

function filterByTags(value: ComponentInfo, terms: string[]): boolean {
  const tags = value.tags?.map((tag) => tag.toLowerCase())
  return terms.some((term) => tags?.includes(term))
}
