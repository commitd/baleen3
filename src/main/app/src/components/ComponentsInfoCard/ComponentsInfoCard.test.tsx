/*-
 * #%L
 * Baleen 3
 * %%
 * Copyright (C) 2020 Dstl
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import React from 'react'
import { ComponentsInfoCard } from '.'
import { ComponentInfo } from '../../types'
import { render, renderDark, userEvent } from '../../utils/test'

const realInfo: ComponentInfo = {
  artifact: 'annot8-components-vehicles',
  componentClass: 'io.annot8.components.vehicles.processors.GenericVehicle',
  description: 'Extracts vehicles (with descriptions) from text',
  name: 'Generic Vehicle',
}

const fullInfo: ComponentInfo = {
  artifact: 'string',
  componentClass: 'string',
  description: 'string',
  name: 'string',
  settingsClass: 'string',
  tags: ['tag'],
}
it('renders without error', () => {
  const { asFragment } = render(<ComponentsInfoCard info={fullInfo} />)
  expect(asFragment()).toMatchSnapshot()
})

it('renders dark without error', () => {
  const { asFragment } = renderDark(<ComponentsInfoCard info={fullInfo} />)
  expect(asFragment()).toMatchSnapshot()
})

it('renders with partial information', () => {
  const { asFragment } = renderDark(<ComponentsInfoCard info={realInfo} />)
  expect(asFragment()).toMatchSnapshot()
})

it('renders with selected', () => {
  const { asFragment } = render(
    <ComponentsInfoCard selectable selected info={realInfo} />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('renders with selected', () => {
  const toggleSelected = jest.fn()
  const { getByRole } = render(
    <ComponentsInfoCard
      selectable
      selected
      toggleSelected={toggleSelected}
      info={realInfo}
    />
  )
  userEvent.click(getByRole('button'))
  expect(toggleSelected).toHaveBeenCalled()
})
