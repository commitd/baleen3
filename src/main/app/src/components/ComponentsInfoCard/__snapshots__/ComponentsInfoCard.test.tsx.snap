// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders dark without error 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root Styled(WithStyles(ForwardRef(Card)))-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
        >
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                string
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                string
              </span>
            </div>
          </div>
          <p
            class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
          >
            string
          </p>
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <div
              class="MuiChip-root WithStyles(ForwardRef(Chip))-root WithStyles(ForwardRef(Chip))-root MuiChip-sizeSmall"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                tag
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with partial information 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root Styled(WithStyles(ForwardRef(Card)))-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
        >
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                Generic Vehicle
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                io.annot8.components.vehicles.processors.GenericVehicle
              </span>
            </div>
          </div>
          <p
            class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
          >
            Extracts vehicles (with descriptions) from text
          </p>
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders with selected 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root Styled(WithStyles(ForwardRef(Card)))-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <button
          class="MuiButtonBase-root MuiCardActionArea-root"
          tabindex="0"
          type="button"
        >
          <div
            class="Component-root Component-root Styled(Component)-root Styled(Component)-root"
          />
          <div
            class="Component-root Component-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41L9 16.17z"
              />
            </svg>
          </div>
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <div
              class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
            >
              <div
                class="MuiCardHeader-content"
              >
                <span
                  class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
                >
                  Generic Vehicle
                </span>
                <span
                  class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
                >
                  io.annot8.components.vehicles.processors.GenericVehicle
                </span>
              </div>
            </div>
            <p
              class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
            >
              Extracts vehicles (with descriptions) from text
            </p>
            <div
              class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
            />
          </div>
          <span
            class="MuiCardActionArea-focusHighlight"
          />
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders without error 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
    >
      <div
        class="MuiPaper-root MuiCard-root WithStyles(ForwardRef(Card))-root WithStyles(ForwardRef(Card))-root Styled(WithStyles(ForwardRef(Card)))-root MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
        >
          <div
            class="MuiCardHeader-root WithStyles(ForwardRef(CardHeader))-root WithStyles(ForwardRef(CardHeader))-root"
          >
            <div
              class="MuiCardHeader-content"
            >
              <span
                class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
              >
                string
              </span>
              <span
                class="MuiTypography-root MuiCardHeader-subheader MuiTypography-caption MuiTypography-colorTextSecondary MuiTypography-displayBlock"
              >
                string
              </span>
            </div>
          </div>
          <p
            class="MuiTypography-root WithStyles(ForwardRef(Typography))-root WithStyles(ForwardRef(Typography))-root WrappedTypography-root WrappedTypography-root MuiTypography-body1"
          >
            string
          </p>
          <div
            class="MuiBox-root MuiBox-root Styled(MuiBox)-root Styled(MuiBox)-root"
          >
            <div
              class="MuiChip-root WithStyles(ForwardRef(Chip))-root WithStyles(ForwardRef(Chip))-root MuiChip-sizeSmall"
            >
              <span
                class="MuiChip-label MuiChip-labelSmall"
              >
                tag
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
